internal/controller/factory/statefulset.go

```diff
@@ -221,3 +194,65 @@ func generateEtcdCommand(cluster *etcdaenixiov1alpha1.EtcdCluster) []string {
 
 	return command
 }
+
+func getStartupProbe(probe *corev1.Probe) *corev1.Probe {
+	defaultProbe := &corev1.Probe{
+		ProbeHandler: corev1.ProbeHandler{
+			HTTPGet: &corev1.HTTPGetAction{
+				Path: "/readyz?serializable=false",
+				Port: intstr.FromInt32(2379),
+			},
+		},
+		InitialDelaySeconds: 1,
+		PeriodSeconds:       5,
+	}
+	return mergeWithDefaultProbe(probe, defaultProbe)
+}
+
+func getReadinessProbe(probe *corev1.Probe) *corev1.Probe {
+	defaultProbe := &corev1.Probe{
+		ProbeHandler: corev1.ProbeHandler{
+			HTTPGet: &corev1.HTTPGetAction{
+				Path: "/readyz",
+				Port: intstr.FromInt32(2379),
+			},
+		},
+		InitialDelaySeconds: 5,
+		PeriodSeconds:       5,
+	}
+	return mergeWithDefaultProbe(probe, defaultProbe)
+}
+
+func getLivenessProbe(probe *corev1.Probe) *corev1.Probe {
+	defaultProbe := &corev1.Probe{
+		ProbeHandler: corev1.ProbeHandler{
+			HTTPGet: &corev1.HTTPGetAction{
+				Path: "/livez",
+				Port: intstr.FromInt32(2379),
+			},
+		},
+		InitialDelaySeconds: 5,
+		PeriodSeconds:       5,
+	}
+	return mergeWithDefaultProbe(probe, defaultProbe)
+}
+
+func mergeWithDefaultProbe(probe *corev1.Probe, defaultProbe *corev1.Probe) *corev1.Probe {
+	if probe == nil {
+		return defaultProbe
+	}
+
+	if probe.InitialDelaySeconds != 0 {
+		defaultProbe.InitialDelaySeconds = probe.InitialDelaySeconds
+	}
+
+	if probe.PeriodSeconds != 0 {
+		defaultProbe.PeriodSeconds = probe.PeriodSeconds
+	}
+
+	if probe.HTTPGet != nil {
```

Comment:
Probes can be of other types and they're mutually exclusive. I think here should be a check for any of them in `probe`

Reply:
I think it is rather good provided that user can customize image and add custom scripts for health checks

Reply:
Do you think we also need to validate if there is more than one probe?
> Probes can be of other types and they're mutually exclusive. I think here should be a check for any of them in probe

Reply:
I think yes, it should be "one of"

Reply:
> I think yes, it should be "one of"

@sircthulhu  But k8s resources doesnt validate it, and other operators too, not sure do we want to do this?

Reply:
if kubernetes does not validate it, we can leave it as is